http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/AwsCrtResponseBodyPublisher.java

```diff
@@ -0,0 +1,285 @@
+
+/**
+ * Adapts an AWS Common Runtime Response Body stream from CrtHttpStreamHandler to a Publisher<ByteBuffer>
+ */
+@SdkInternalApi
+public class AwsCrtResponseBodyPublisher implements Publisher<ByteBuffer> {
+    private static final Logger log = Logger.loggerFor(AwsCrtResponseBodyPublisher.class);
+    private static final LongUnaryOperator DECREMENT_IF_GREATER_THAN_ZERO = x -> ((x > 0) ? (x - 1) : (x));
+
+    private final HttpClientConnection connection;
+    private final HttpStream stream;
+    private final CompletableFuture<Void> responseComplete;
+    private final AtomicLong outstandingRequests = new AtomicLong(0);
+    private final int windowSize;
+    private final AtomicBoolean isCancelled = new AtomicBoolean(false);
+    private final AtomicBoolean areNativeResourcesReleased = new AtomicBoolean(false);
+    private final AtomicBoolean isSubscriptionComplete = new AtomicBoolean(false);
+    private final AtomicBoolean queueComplete = new AtomicBoolean(false);
+    private final AtomicInteger mutualRecursionDepth = new AtomicInteger(0);
+    private final AtomicInteger queuedBytes = new AtomicInteger(0);
+    private final AtomicReference<Subscriber<? super ByteBuffer>> subscriberRef = new AtomicReference<>(null);
+    private final Queue<byte[]> queuedBuffers = new ConcurrentLinkedQueue<>();
+    private final AtomicReference<Throwable> error = new AtomicReference<>(null);
+
+    /**
+     * Adapts a streaming AWS CRT Http Response Body to a Publisher<ByteBuffer>
+     * @param stream The AWS CRT Http Stream for this Response
+     * @param windowSize The max allowed bytes to be queued. The sum of the sizes of all queued ByteBuffers should
+     *                   never exceed this value.
+     */
+    public AwsCrtResponseBodyPublisher(HttpClientConnection connection, HttpStream stream,
+                                       CompletableFuture<Void> responseComplete, int windowSize) {
+        Validate.notNull(connection, "HttpConnection must not be null");
+        Validate.notNull(stream, "Stream must not be null");
+        Validate.notNull(responseComplete, "Stream must not be null");
+        Validate.isPositive(windowSize, "windowSize must be > 0");
+        this.connection = connection;
+        this.stream = stream;
+        this.responseComplete = responseComplete;
+        this.windowSize = windowSize;
+    }
+
+    /**
+     * Method for the users consuming the Http Response Body to register a subscriber.
+     * @param subscriber The Subscriber to register.
+     */
+    @Override
+    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
+        Validate.notNull(subscriber, "Subscriber must not be null");
+
+        boolean wasFirstSubscriber = subscriberRef.compareAndSet(null, subscriber);
+
+        if (!wasFirstSubscriber) {
+            log.error(() -> "Only one subscriber allowed");
+            subscriber.onError(new IllegalStateException("Only one subscriber allowed"));
```

Comment:
Unsure of the right way to proceed here.  We only want one subscriber so should we just drop the onError signal and return, or drop it and throw the exception that was being pushed?  The latter seems best, but I want to check first.

Reply:
You can still signal `onError` for all subscribers besides the first one. The spec just requires that `onSubscribe` is called before signalling anything. The typical way to handle this would be to simply call `onSubscribe` with a dummy `Subscription`:

```java
if (!wasFirstSubscriber) {
    subscriber.onSubscribe(new DummySubscription());
    subscriber.onError(...);
}
```

This mentioned in the doc:

> ...for all other situations the only legal way to signal failure (or reject the Subscriber) is by calling onError (after calling onSubscribe).